src/main/scala/sangria/relay/Connection.scala

```diff
@@ -127,36 +126,64 @@ object Connection {
       if (l < 0)
         throw ConnectionArgumentValidationError("Argument 'last' must be a non-negative integer"))
 
-    val sliceEnd = sliceStart + seqSlice.size
-    val beforeOffset = getOffset(before, size)
-    val afterOffset = getOffset(after, -1)
+    val SliceInfo(sliceStart, arrayLength) = sliceInfo
+
+    val sliceEnd = sliceStart + arraySlice.size
 
-    val startOffset = math.max(math.max(sliceStart - 1, afterOffset), -1) + 1
-    val endOffset = math.min(math.min(sliceEnd, beforeOffset), size)
+    val startOffset = math.max(sliceStart, 0)
+    val endOffset = math.min(sliceEnd, arrayLength)
 
-    val actualEndOffset = first.fold(endOffset)(f => math.min(endOffset, startOffset + f))
-    val actualStartOffset = last.fold(startOffset)(l => math.max(startOffset, actualEndOffset - l))
+    val afterOffset = getOffset(after, -1)
+    val (firstEdgeOffset, actualStartOffset) = if (afterOffset >= 0 && afterOffset < arrayLength) {
+      val actualStartOffset = math.max(startOffset, afterOffset + 1)
+      val fEO = afterOffset + 1
+      fEO -> actualStartOffset
+    } else 0 -> startOffset
+
+    val beforeOffset = getOffset(before, arrayLength)
+    val (lastEdgeOffset, actualEndOffset) = if (0 <= beforeOffset && beforeOffset < arrayLength) {
+      val lEO = beforeOffset - 1
+      val actualEndOffset = math.min(endOffset, beforeOffset)
+      lEO -> actualEndOffset
+    } else (arrayLength - 1) -> endOffset
+
+    val numberEdgesAfterCursor = lastEdgeOffset - firstEdgeOffset + 1
+
+    val finalEndOffset =
+      first.map(_ + actualStartOffset).map(math.min(actualEndOffset, _)).getOrElse(actualEndOffset)
+    val finalStartOffset =
+      last.map(actualEndOffset - _).map(math.max(actualStartOffset, _)).getOrElse(actualStartOffset)
```

Comment:
minor: we can use `fold` to avoid some function calls.
```suggestion
    val finalEndOffset =
      first.fold(actualEndOffset)(f => math.min(actualEndOffset, f + actualStartOffset))
    val finalStartOffset =
      last.fold(actualStartOffset)(l => math.max(actualStartOffset, actualEndOffset - l))
```